### ISSUE TITLE
go_test rule doesn't support TestMain

### ISSUE THREAD

**Issue by adrianludwin** [bazelbuild/rules_go#64]
I should be able to create a function with the signature:

`func TestMain(m *testing.M)`

which is used to setup/teardown common work before/after each test. But `go/tools/generate_test_main.go` includes this function as one of the tests, which is incorrect. 

The following patch _appears_ to fix the problem:

~~~~
diff --git a/go/tools/generate_test_main.go b/go/tools/generate_test_main.go
index b78f124..29fc4b9 100644
--- a/go/tools/generate_test_main.go
+++ b/go/tools/generate_test_main.go
@@ -76,6 +76,10 @@ func main() {
                        if !strings.HasPrefix(fn.Name.Name, "Test") {
                                continue
                        }
+                       if fn.Name.Name == "TestMain" {
+                               // TestMain is not, itself, a test
+                               continue
+                       }
                        cases.Names = append(cases.Names, fn.Name.Name)
                }
        }
~~~~

But I don't see any tests for this rule so I'm not sure how to verify this change. 

CC @yugui - I hear you're the guru on these rules :) Also CCing @kchodorow - are you the owner of this file/repo?

**Comment by adrianludwin** [comment]
A little more investigation showed that TestMain wasn't even being called. This appears to fix that problem:
~~~~
@@ -30,9 +30,10 @@ import (
 
 // Cases holds template data.
 type Cases struct {
-       Package string
-       RunDir  string
-       Names   []string
+       Package     string
+       RunDir      string
+       Names       []string
+       HasTestMain bool
 }
 
 func main() {
@@ -76,6 +77,11 @@ func main() {
                        if !strings.HasPrefix(fn.Name.Name, "Test") {
                                continue
                        }
+                       if fn.Name.Name == "TestMain" {
+                               // TestMain is not, itself, a test
+                               cases.HasTestMain = true
+                               continue
+                       }
                        cases.Names = append(cases.Names, fn.Name.Name)
                }
        }
@@ -101,7 +107,12 @@ var tests = []testing.InternalTest{
 
 func main() {
   os.Chdir("{{.RunDir}}")
+  {{if not .HasTestMain}}
   testing.Main(everything, tests, nil, nil)
+  {{else}}
+  m := testing.MainStart(everything, tests, nil, nil)
+  undertest.TestMain(m)
+  {{end}}
 }
 `))
        if err := tpl.Execute(outFile, &cases); err != nil {
~~~~

However, it is hacky. It would be nice if we could find a better way of generating this file.

**Comment by adrianludwin** [comment]
Actually checked change into a personal repo: https://github.com/adrianludwin/rules_go/commit/0ff84610156d26cd1f9836dd8034c29a82baac5d

**Comment by yugui** [comment]
@adrianludwin Right. The go toolchain checks the signature of the function a bit more strictly, but I think your implementation is good enough for us.

Would you mind sending a pull request for your change?

FYI, the issue itself is a duplicate of #11.

**Comment by yugui** [comment]
Fixed with #68.

@adrianludwin Thank you for your contribution!
